ERCS/erc-7662.md

```diff
@@ -0,0 +1,103 @@
+---
+eip: 7662
+title: AI Agent NFTs
+description: A specification for NFTs that represent AI Agents.
+author: Greg Marlin (@0marleymarl)
+discussions-to: https://ethereum-magicians.org/t/erc-7662-ai-agent-nfts/19371
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-03-26
+requires: 721
+---
+
+## Abstract
+
+This proposal introduces a standard for AI agent NFTs. In order for AI Agents to be created and traded as NFTs it doesn't make sense to put the prompts in the token metadata, therefore it requires a standard custom struct. It also needs doesn't make sense to store the prompts directly onchain as they can be quite large, therefore this standard proposes they be stored as decentralized storage URLs. This standard also proposes two options on how this data should be made private to the owner of the NFT, with the favored implementation option being encrypting the data using custom contract parameters for decryption that decrypt only to the owner of the NFT. 
+
+## Motivation
+
+The creation and trading of AI Agent NFTs are a natural fit and offer the potential for an entirely new onchain market. This requires some custom data to be embedded in the NFT through a custom struct and this needs to be standardized so that any marketplace or AI Agent management product, among others, know how to create and parse AI Agent NFTs.  
+
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+
+All ERC-XXXX compliant contracts MUST implement the standard [ERC-721](./eip-721.md) functionality for minting and transferring NFTs, and MUST additionally implement this standard's Agent struct
+
+```solidity
+   
+struct Agent {
```

Comment:
Can you elaborate why this needs to be stored onchain as opposed to be stored as ERC721 Metadata?

Reply:
The primary reason for not storing them directly in the metadata is for the case if someone didn’t encrypt the prompts. If they are stored onchain someone could still discover them but it would be a lot harder for a casual user, for example looking at the NFT on a marketplace. Many marketplaces list a link to the TokenURI that can easily be looked up, and thus the casual user could easily just grab the prompts and not buy the NFT, whereas if it's stored onchain they can't. If stored onchain unencrypted, a determined user could e.g. look up the info on a block explorer or look it up from a self-built app but considering most of these I expect to be priced fairly low then it’s easier to just pay for it, for most cases. For high-value agents, then those sellers should obviously opt for the encryption.

Since I’m recommending encryption of the prompts anyway, and will be using that in the reference implementation, it could make sense to adapt the Metadata Standard for this since encrypted prompt URIs could be exposed in the metadata, however I’ve tried to avoid enforcing the need for encryption in the standard if that makes sense. 

Another consideration for putting it onchain is to be a foundation for further future extensibility in terms of building out new data features to that struct (that can be accessed with the same call i.e. Agents[id]) and opening up the possibility of other smart contracts interacting with the data in that struct.